src/math/ring_signature.cpp

```diff
@@ -0,0 +1,363 @@
+    auto secret_keys = generate_keys_map(secrets);
+    // Organize known public keys into corresponding rings
+    const auto known_keys_by_ring =
+        partition_keys_into_rings(secret_keys, rings);
+    // Check we know a secret key in each ring
+    if (!all_rings_nonzero_size(known_keys_by_ring))
+        return false;
+
+    // Compute indexes for known keys inside the rings.
+    const index_list known_key_indexes =
+        search_key_indexes(rings, known_keys_by_ring);
+    BITCOIN_ASSERT(known_key_indexes.size() == rings.size());
+
+    // Compute message digest M
+    const data_chunk message_data = concatenate(message, rings);
+    const hash_digest M = sha256_hash(message_data);
+
+    // Generate random k and s values
+    secret_list k;
+    random_k_and_s_values(k, out.s, rings, seed);
+
+    // ---------------------------------------------------------------------
+    // Step 1: calculate e0
+    // ---------------------------------------------------------------------
+    data_chunk e0_data;
+
+    for (size_t i = 0; i < rings.size(); ++i)
+    {
+        BITCOIN_ASSERT(i < known_key_indexes.size());
+        BITCOIN_ASSERT(i < k.size());
+        BITCOIN_ASSERT(i < out.s.size());
+
+        // Current ring and index of known key
+        const auto& ring = rings[i];
+        const size_t known_key_index = known_key_indexes[i];
+
+        // Calculate starting R value...
+        ec_compressed R_i_j;
+        bool rc = secret_to_public(R_i_j, k[i]);
+
+        // ... Start one above index of known key and loop until the end
+        for (size_t j = known_key_index + 1; j < ring.size(); ++j)
+        {
+            BITCOIN_ASSERT(j < out.s[i].size());
+            const auto& s = out.s[i][j];
+
+            // Calculate e and R until the end of this ring.
+            const auto e_i_j = borromean_hash(M, R_i_j, i, j);
+            R_i_j = calculate_R(s, e_i_j, ring[j]);
+        }
+        // Add this ring to e0
+        extend_data(e0_data, R_i_j);
+    }
+    extend_data(e0_data, M);
+    // Hash data to produce e0 value
+    out.e = sha256_hash(e0_data);
+
+    // ---------------------------------------------------------------------
+    // Step 2: join up each ring at the index where we know the secret key
+    // ---------------------------------------------------------------------
+    for (size_t i = 0; i < rings.size(); ++i)
+    {
+        BITCOIN_ASSERT(i < known_key_indexes.size());
+        BITCOIN_ASSERT(i < k.size());
+        BITCOIN_ASSERT(i < out.s.size());
+
+        // Current ring and index of known key
+        const auto& ring = rings[i];
+        const size_t known_key_index = known_key_indexes[i];
+
+        // Calculate starting e value of this current ring.
+        auto e_i_j = borromean_hash(M, out.e, i, 0);
+
+        // Loop until index of known key.
+        for (size_t j = 0; j < known_key_index; ++j)
+        {
+            BITCOIN_ASSERT(j < out.s[i].size());
```

Comment:
Test outside of loop. Is this assertion *possible*?

Reply:
done